packages/gatsby/src/query/query-runner.ts

```diff
@@ -172,11 +176,11 @@ export async function queryRunner(
     .digest(`base64`)
 
   if (
-    resultHash !== resultHashes.get(queryJob.id) ||
+    resultHash !== (await resultHashCache.get(queryJob.id)) ||
     (queryJob.isPage &&
       !pageDataExists(path.join(program.directory, `public`), queryJob.id))
   ) {
-    resultHashes.set(queryJob.id, resultHash)
+    resultHashCache.set(queryJob.id, resultHash)
```

Comment:
Doesn't this cache only help us on the next build? Correct me if I'm wrong but isn't queryJob.id = page path? Which leads to uniqueness per build? So I'm not sure if this change does anything for us, unless my above assumption is wrong.

If I'm wrong above, this solution doesn't guard us 100% because the cache is async which can lead to two workers still going through at the same time, to mitigate this we use a mutex or we add a comment that we're aware and fine with it.

Reply:
> If I'm wrong above, this solution doesn't guard us 100% because the cache is async which can lead to two workers still going through at the same time, to mitigate this we use a mutex or we add a comment that we're aware and fine with it.

Generally yes, but Gatsby doesn't "randomly" run queries and especially not 2 of the "same" queries (by path) at the same time. We probably should put `await` before `resultHashCache.set`, but we shouldn't need mutex for it

Reply:
Oh my bad I didn't see it was a Map before, I thought we're using cache-fs 🤦 Now this makes more sense!